src/store_tests.cc

```diff
@@ -410,3 +217,162 @@ bool test_init_and_recover_containers(bool print_all) {
   return true;
 }
 
+bool test_policy_store(bool print_all) {
+  policy_store ps(500);
+
+  key_message pk;
+  if (!make_certifier_rsa_key(2048,  &pk))
+    return false;
+
+  time_point t_nb;
+  if (!time_now(&t_nb))
+    return false;
+  string s_nb;
+  if (!time_to_string(t_nb, &s_nb))
+    return false;
+
+  double hours_to_add = 365.0 * 24.0;
+  time_point t_na;
+  if (!add_interval_to_time_point(t_nb, hours_to_add, &t_na))
+    return false;
+  string s_na;
+  if (!time_to_string(t_na, &s_na))
+    return false;
+
+  pk.set_not_before(s_nb);
+  pk.set_not_after(s_na);
+
+  if (!ps.set_policy_key(pk))
+    return false;
+
+  const key_message* pkt = ps.get_policy_key();
+  if (pkt == nullptr)
+    return false;
+  if (!same_key(pk, *pkt))
+    return false;
+  if (print_all) {
+    printf("policy-key store/retrieve works\n");
+  }
+
+  if (ps.get_num_entries() != 0) {
+    printf("Error: policy-key store number of entries should be 0, but is %d\n",
+        ps.get_num_entries());
+    return false;
+  }
+
+  byte bin[5] = {
+    0, 1, 2, 3,4
+  };
+  string tag1;
+  string type1;
+  string value1;
+
+  tag1 = "test-entry-1";
+  type1 = "binary";
+  value1.assign((const char*)bin, sizeof(bin));
+  if (!ps.add_entry(tag1, type1, value1)) {
+    printf("Error:Can't add entry\n");
+    return false;
+  }
+
+  tag1 = "test-entry-2";
+  type1 = "binary";
+  value1.assign((const char*)bin, sizeof(bin));
+  if (!ps.add_entry(tag1, type1, value1)) {
+    printf("Error:Can't add entry\n");
+    return false;
+  }
+
+  tag1 = "test-entry-3";
+  type1 = "binary";
+  value1.assign((const char*)bin, sizeof(bin));
+  if (!ps.update_or_insert(tag1, type1, value1)) {
+    printf("Error: Can't update or insert entry\n");
+    return false;
+  }
+
+  if (ps.get_num_entries() != 3) {
+    printf("Error: policy-key store number of entries is %d should be 3\n", ps.get_num_entries());
+    return false;
+  }
+
+  tag1 = "test-entry-2";
+  type1 = "binary";
+  int ent = ps.find_entry_by_tag(tag1);
```

Comment:
The business of uniqueness on either `(tag)` or on the pair `(tag, type)`, still bothers me.

What if someone adds another entry with same tag: 'test-entry-2'. Then this find() can never return the 2nd entry, till the 1st entry is deleted.

I think this make the interface a bit clunky and under-specified in its usefulness.

Still waiting to hear how you plan to reconcile these uniqueness semantics.